### PR TITLE
Feat/Payment Integration with Multiple Payment Gateways

### PR DESCRIPTION
**Pull request by ZainabAlayande**
## How this should be manually

**Try accessing the endpoint**

- Access routes as an authenticated user.
- Send a POST request to `/api/v1/payments/flutterwave ` or `/api/v1/payments/paystack ` to subscribe to pricing plan with either of the configured payment gateways.

#### Request
```
{
    "organizationId": "4ae0a02c-b595-49ac-8e95-763adbde2c53",
    "planId": "2",
    "fullName": "John Doe",
    "billingOption": "monthly",
    "redirectUrl": "https://api-java.boilerplate.hng.tech/docs"
}

```

#### Response
```
{
    "status": "200",
    "message": "Payment Initiated Successfully",
    "data": {
        "payment_url": "https://checkout.paystack.com/9xnm7ez2ax8klcv"
    }
}
```

## Checklist of What I did

- [ ] Integrated paystack subscription API to enable users subscribe to the pricing plan of choice
- [ ] Integrated flutterwave subscription API to enable users subscribe to the pricing plan of choice
- [ ] Implemented error handling and returned appropriate status codes.
- [ ] Securely stored organization subscription information in the database.
- [ ] Wrote unit test to test the service logic.

## Reference Issue
Link to Issue: [https://github.com/hngprojects/hng_boilerplate_java_web/issues/267]()

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/payment/controller/PaymentController.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    private final PaymentService paymentService;
+
+    private final UserService userService;
+
+    private PaymentServiceFactory paymentServiceFactory;
+
+
+    @Autowired
+    public PaymentController(@Qualifier("paystackService") PaymentService paymentService, UserService userService, PaymentServiceFactory paymentServiceFactory) {
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/payment/controller/PaymentController.java:26]: you don't need to annotate constructor injection with autowired

File: src/main/java/hng_java_boilerplate/payment/service/payment/FlutterwaveServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    private final UserService userService;
+
+    private final PaymentRepository paymentRepository;
+    private final Utils utils;
+
+    @Value("${flutterwave.secret.key}")
+    private String flutterwaveSecretKey;
+
+//    @Autowired
```
Comment (Fifetoyi) [inline src/main/java/hng_java_boilerplate/payment/service/payment/FlutterwaveServiceImpl.java:38]: remove comment

File: src/main/resources/db/migration/V13__add_two_fa_and_secret_key_col.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
+-- Add columns to the users table
+ALTER TABLE users
+    ADD COLUMN two_FA_status BOOLEAN,
+ADD COLUMN two_FA_secret_key VARCHAR(255);
+
+
+-- V6__add_two_fa_columns_and_backup_codes_table.sql
+-- Create the two_fa_backup_codes table
+CREATE TABLE two_fa_backup_codes (
+                                     user_id VARCHAR(255) NOT NULL,
```
Comment (Fifetoyi) [inline src/main/resources/db/migration/V13__add_two_fa_and_secret_key_col.sql:10]: format properly

File: src/main/java/hng_java_boilerplate/payment/service/payment/PaystackServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            dataObject.put("paid_at", "");
+        }
+        PaymentVerificationResponse verificationResponse = PaymentVerificationResponse.builder().message("Verification Successful").status_code("200").data(data).build();
+        return ResponseEntity.ok(verificationResponse);
+    }
+
+
+    @Override
+    public ResponseEntity<?> verifyPayment(String reference, String status, String transactionId) {
+        return null;
```
Comment (Fifetoyi) [inline src/main/java/hng_java_boilerplate/payment/service/payment/PaystackServiceImpl.java:121]: is this method used? should it return null?

File: src/main/java/hng_java_boilerplate/payment/Utils.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    public String generateTransactionReference() {
+        return "TRANS_" + UUID.randomUUID();
+    }
+
+    public User validateLoggedInUser() {
+        User user = userService.getLoggedInUser();
+        if (user != null) {
+            return user;
+        } else {
+            throw new UserNotFoundException("User not authorized");
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/payment/Utils.java:31]: shouldn't this be unauthorized exception 401?

Comment (lurldgbodex) [reply src/main/java/hng_java_boilerplate/payment/Utils.java:31]: The method is already throwing a UserNotFound; The if else is redunctant and if user is null will never run

File: src/main/java/hng_java_boilerplate/payment/controller/UserPaymentController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+
+    private final UserPaymentService paymentService;
+
+    private final UserService userService;
+
+
+
+    @Autowired
+    public UserPaymentController(UserPaymentService paymentService, UserService userService) {
```
Comment (lurldgbodex) [inline src/main/java/hng_java_boilerplate/payment/controller/UserPaymentController.java:28]: is autowired needed for constructor injection?

**Pull request closed without merging**